pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -39,15 +64,39 @@ type provider interface {
 type Service struct {
 	outboundTransport transport.OutboundTransport
 	store             storage.Store
+	callbackChannel   chan didCommChMessage
+	actionEvent       chan<- event.DIDCommEvent
+	streamEvents      []chan<- dispatcher.DIDCommMsg
+	lock              sync.RWMutex
+	streamLock        sync.RWMutex
+	execute           bool
 }
 
 // New return didexchange service
 func New(store storage.Store, prov provider) *Service {
-	return &Service{outboundTransport: prov.OutboundTransport(), store: store}
+	svc := &Service{
+		// TODO Outbound dispatcher - https://github.com/hyperledger/aries-framework-go/issues/259
+		outboundTransport: prov.OutboundTransport(),
+		store:             store,
+		// TODO channel size - https://github.com/hyperledger/aries-framework-go/issues/246
+		callbackChannel: make(chan didCommChMessage, 10),
+		// set execute to false. Consumers have to enable this by setting either RegisterActionEvent() or
+		// RegisterAutoExecute()
+		execute: false,
+	}
+
+	svc.startInternalListener()
+
+	return svc
 }
 
 // Handle didexchange msg
 func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
+	// throw error if there are no action events are registered or auto execute set (if it's not outbound)
+	if !msg.Outbound && !s.execute {
```

Comment:
Not sure if this is related to the comment by Troy below, but s.execute needs an RLock in order to read.

Reply:
[Added](https://github.com/hyperledger/aries-framework-go/compare/6dc9c7de378434ea27520364556f649ccf434c5f..36d8a92bfaa82b830e44a047521b376920c82c72#diff-f761c91e753eba1c42a68076df06ce12R97).